File: src/Scanner.java
package tsundere;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static tsundere.TokenType.*;

class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0; // points to the first character of the lexeme
    private int current = 0; // points to the current character of the lexeme
    private int line = 1;

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (current < source.length()) {
            // We are at the beginning of the next lexeme.
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        char c = source.charAt(current++);
        switch (c) {
            case '(': addToken(LEFT_PAREN,  null); break;
            case ')': addToken(RIGHT_PAREN, null); break;
            case '{': addToken(LEFT_BRACE,  null); break;
            case '}': addToken(RIGHT_BRACE, null); break;
            case ',': addToken(COMMA,       null); break;
            case '.': addToken(DOT,         null); break;
            case '-': addToken(MINUS,       null); break;
            case '+': addToken(PLUS,        null); break;
            case ';': addToken(SEMICOLON,   null); break;
            case '*': addToken(STAR,        null); break;
            case '!': addToken(match('=') ? BANG_EQUAL    : BANG,    null); break;
            case '=': addToken(match('=') ? EQUAL_EQUAL   : EQUAL,   null); break;
            case '<': addToken(match('=') ? LESS_EQUAL    : LESS,    null); break;
            case '>': addToken(match('=') ? GREATER_EQUAL : GREATER, null); break;
            case '/':
                if (match('/')) {
                    // keep going until the end of the line
                    while (peek() != '\n' && current < source.length) current++;
                } else {
                    addToken(SLASH, null);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
              // Ignore whitespace.
              break;
            case '\n': line++;  break;
            case '"': 
                while (peek() != '"' && current < source.length) {
                    if (peek() == '\n') line++;
                    current++;
                }

                if (current >= source.length) {
                    Lox.Error("Unterminated string");
                    return;
                }

                // The closing ".
                current++;

                String value = source.substring(start + 1, current - 1);
                addToken(STRING, value);
                break;
            default:
                if (isDigit(c)) {
                    while (isDigit(peek())) current++;

                    // Look for a fractional part
                    if (peek() == '.' && isDigit(peekNext())) {
                        current++;
                        while (isDigit(peek())) current++;
                    }
                    addToken(TOKEN, Double.parseDouble(source.substring(start, current)));
                } else {
                    Lox.error(line, "Unexpected character.");
                }
                break;
        }
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean match(char expected) {
        if (current >= source.length) return false;
        if (source.charAt(current) != expected) return false;
        current++;
        return true;
    }

    private char peek() {
        if (current >= source.length) return '\0';
        return source.charAt(current);
    }

    private void addToken(TokenType type, Object literal) {
        String lexeme = source.substring(start, current);
        tokens.add(new Token(type, lexeme, literal, line));
    }
}

File: src/scanner.cpp
#include "scanner.h"

Scanner::Scanner(std::string source) {
    this->source = source;
}

bool Scanner::is_at_end() { return current >= source.size(); }

std::vector<Token> Scanner::scan_tokens() {
    while (!is_at_end()) {
        start = current;
        scan_token();
    }

    tokens.push_back(Token(EOFF, "", line, 0));
    return tokens;
}

void Scanner::scan_token() {
    switch (source[current++]) {
        case '(': add_token(LEFT_PAREN); break;
        case ')': add_token(RIGHT_PAREN); break;
        case '{': add_token(LEFT_BRACE); break;
        case '}': add_token(RIGHT_BRACE); break;
        case ',': add_token(COMMA); break;
        case '.': add_token(DOT); break;
        case '-': add_token(MINUS); break;
        case '+': add_token(PLUS); break;
        case ';': add_token(SEMICOLON); break;
        case '*': add_token(STAR); break; 
        case '!': add_token(match('=') ? BANG_EQUAL : BANG); break;
        case '=': add_token(match('=') ? EQUAL_EQUAL : EQUAL); break;
        case '<': add_token(match('=') ? LESS_EQUAL : LESS); break;
        case '>': add_token(match('=') ? GREATER_EQUAL : GREATER); break;
        case '/':
            if (match('/')) {
                while (peek() != '\n' && !is_at_end()) current++;
            } else {
                add_token(SLASH);
            }
            break;
        // ingnore white spaces
        case ' ':
        case '\t':
        case '\r':
            break;
        case '\n': line++; break;
        case '"':
            while (peek() != '"' && !is_at_end()) current++;
            if (is_at_end()) {
                std::cout << "Unfinished string" << std::endl;
                exit(1);
            }
            break;
        default:
            std::cout << line << " Unexpected Character" << std::endl;
            break;
    }
}

char Scanner::peek() {
    if (is_at_end()) return '\0';
    return source[current];
}

bool Scanner::match(char expected) {
    if (is_at_end()) return false;
    if (source[current] != expected) return false;
    current++;
    return true;
}

void Scanner::add_token(Token_Type type) {
    add_token(type, 0);
}

void Scanner::add_token(Token_Type type, uint64_t literal) {
    std::string lexeme = source.substr(start, current-start);
    tokens.push_back(Token(type, lexeme, line, literal));
}
